[fasterize/FasterizeEngine#1483] Ensure that the inserted script has valid type

abarre opened the pull request:
jquery mechanism to insert script node implies the usage of a node element with a disabled type (false/ or true/application/javascript).
This kind of nodes is not interpreted by the browsers and will never emit a load event.

dax reviewed (approved):
I think @zimski should also review it. Not sure to understand all the consequences of this change.

Do you think we could imagine a way to execute test automatically?
I think we could then add more tests (not specifically for this case but globally) and it would help us refactor this code easily.

What do you think about testing if suspected dead code is still alive?

zimski commented:
I am on it

abarre commented:
@dax, yes, we could do automatic testing with Karma on different browsers. We could add an issue on this. However, I don't think that this is a priority right now.

dax commented:
I don't think this should be a dedicated issue as it will never be a priority.
The other way to do it is to setup just a single test while implementing a bug/feature issue. Migrating the other tests could then be done little by little.

zimski commented on tests/reference/js_defer_onload.html:
```diff
@@ -0,0 +1,20 @@
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/1.10.2/jquery.js"></script>
+    <script type="text/javascript" >
+      window.onload = function () {
+        document.writeln('onload');
```
the `document.writeln` in both browser ( chrome and firefox ) will remove all the content of the body, and the result is:

![image](https://user-images.githubusercontent.com/1662759/30581941-3198c22a-9d22-11e7-9eb0-2646913d471f.png)

it's too aggressive for this tests

zimski commented on tests/reference/js_defer_onload.html:
```diff
@@ -0,0 +1,20 @@
+      node.async = true;
+      node.src = "https://cdnjs.cloudflare.com/ajax/libs/jqueryui/1.12.1/jquery-ui.js";
+
+      $(document).find("body").append(node); // it fails, the script is not inserted but fetch in Ajax !!!
```
When I replace the `document.writeln` by an other instruction like `$('body').append('onload');` everything is working okay, I have my message and the script is correctly inserted.

![image](https://user-images.githubusercontent.com/1662759/30582062-a85b1386-9d22-11e7-8525-307be80809a9.png)

So I think that the `writeln` is executed after the insertion of the tag so it will remove every thing and this is what you have observed by your comment ` it fails, the script is not inserted but fetched in ajax`

zimski reviewed (changes requested):
As I said in my comments the `writeln` has an unwanted behaviour in this tests ( removing the whole content of the body ) and should be replaced by something else less destructif.

abarre committed:
[fasterize/FasterizeEngine#1483] Automate tests

Tests are done against versions with and without deferjs.
Results must be identical between the two versions.

Tests are executed in headless chrome.

abarre committed:
[fasterize/FasterizeEngine#1483] Count only valid inserted scripts

jquery mechanism to insert script node implies the usage of a node element with a disabled type attribute (false/ or true/application/javascript).
This kind of nodes is not interpreted by the browsers and will never emit a load event.

abarre committed:
[fasterize/FasterizeEngine#1483] Remove dead code

- remove all codes related to IE10 and below
- remove code related to BEFORE_SCRIPTS and AFTER_SCRIPTS hooks.

abarre committed:
[fasterize/FasterizeEngine#1483] Fix js minification

- update uglifyjs and activate mangle-props and compress.
- ensure that tests pass on the prod build.

abarre commented:
I updated the PR with :  
- automatic tests
- remove dead code
- fix JS minification

The size of js_defer-min.js drop from 19.9k to 12.2k.

dax commented:
I will let @zimski review the change on the code but this is really cool you have added automated tests :)

zimski commented on package.json:
```diff
@@ -1,18 +1,25 @@
   "scripts": {
-    "build": "./fasterize_build.sh"
+    "build": "./fasterize_build.sh",
+    "dev:test": "mocha ./runner.js tests/test.js",
```
this is not working, maybe you should fix the `runner.js` path

zimski commented on package.json:
```diff
@@ -1,18 +1,25 @@
-    "build": "./fasterize_build.sh"
+    "build": "./fasterize_build.sh",
+    "dev:test": "mocha ./runner.js tests/test.js",
+    "test": "npm run build && DEFER_FILE=js_defer-min.js mocha --timeout 10000 ./tests/utils/runner.js tests/test.js"
```
I think this test should run the `compare_all` tests too.
btw I was unable to run the `compare_all` tests, I get an error about 
```
Error: Cannot find module 'webpage'
    at Function.Module._resolveFilename (module.js:489:15)
    ....

```

zimski reviewed (changes requested):
Great job about adding this automatic tests ! this will help us a lot to maintain the code quality and encourage us to add feature/refactore without having this fear to breaks everything.

abarre commented:
What do you call compare_all ? I don't find any reference of it in the
project.

-- 
*Anthony BARRE*
Software Engineer
+339 72 62 85 68 <+339+72+62+85+68>

Fasterize - Accélérateur de site web

2017-09-25 10:25 GMT+02:00 CHADDA Chakib <notifications@github.com>:

> *@zimski* requested changes on this pull request.
>
> Great job about adding this automatic tests ! this will help us a lot to
> maintain the code quality and encourage us to add feature/refactore without
> having this fear to breaks everything.
>
> I get some little troubles to run the compare_all tests, if you can help
> to figure this out
> ------------------------------
>
> In package.json
> <https://github.com/fasterize/defer.js/pull/6#discussion_r140713137>:
>
> >    "repository": {
>      "type": "git",
>      "url": "git://github.com/fasterize/defer.js.git"
>    },
>    "scripts": {
> -    "build": "./fasterize_build.sh"
> +    "build": "./fasterize_build.sh",
> +    "dev:test": "mocha ./runner.js tests/test.js",
>
> this is not working, maybe you should fix the runner.js path
> ------------------------------
>
> In package.json
> <https://github.com/fasterize/defer.js/pull/6#discussion_r140713638>:
>
> >    "repository": {
>      "type": "git",
>      "url": "git://github.com/fasterize/defer.js.git"
>    },
>    "scripts": {
> -    "build": "./fasterize_build.sh"
> +    "build": "./fasterize_build.sh",
> +    "dev:test": "mocha ./runner.js tests/test.js",
> +    "test": "npm run build && DEFER_FILE=js_defer-min.js mocha --timeout 10000 ./tests/utils/runner.js tests/test.js"
>
> I think this test should run the compare_all tests too.
> btw I was unable to run the compare_all tests, I get an error about
>
> Error: Cannot find module 'webpage'
>     at Function.Module._resolveFilename (module.js:489:15)
>     ....
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/fasterize/defer.js/pull/6#pullrequestreview-64822884>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA4rvbTpml31rcyduckCapQ6V9lCIPOCks5sl2N6gaJpZM4PZAMr>
> .
>

zimski commented:
I am sorry, I remove my previous comment, it's was related to an old code untracked in my repo.
just fix the the `dev:test` and it's OK

abarre committed:
[fasterize/FasterizeEngine#1483] Fix dev:test task.
